File: fasta/Version3/Claude/fasta3-opt.c
/*
 * The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * Converted from Python to C with optimizations by Claude
 * Original Python by Ian Osgood, Heinrich Acker, Justin Peel, and Christopher Sean Forgeron
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <pthread.h>
#include <unistd.h>

#define LINE_WIDTH 60
#define IM 139968.0
#define BUFFER_SIZE (64 * 1024)  // 64KB buffer for output
#define NUM_BUFFERS 2            // Double buffering for output

// ALU sequence
const char *alu =
   "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
   "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
   "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
   "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
   "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
   "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
   "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

// Frequency tables
typedef struct {
    char c;
    double p;
} Frequency;

Frequency iub[] = {
    {'a', 0.27}, {'c', 0.12}, {'g', 0.12}, {'t', 0.27},
    {'B', 0.02}, {'D', 0.02}, {'H', 0.02}, {'K', 0.02},
    {'M', 0.02}, {'N', 0.02}, {'R', 0.02}, {'S', 0.02},
    {'V', 0.02}, {'W', 0.02}, {'Y', 0.02}
};

Frequency homosapiens[] = {
    {'a', 0.3029549426680},
    {'c', 0.1979883004921},
    {'g', 0.1975473066391},
    {'t', 0.3015094502008}
};

// Lookup table for faster character selection
typedef struct {
    double prob;
    char c;
} CumulativeFreq;

// Buffer for asynchronous I/O
typedef struct {
    char data[BUFFER_SIZE];
    size_t size;
    int ready;
    pthread_mutex_t mutex;
} Buffer;

// Globals for double buffering
Buffer buffers[NUM_BUFFERS];
int current_buffer = 0;
int write_buffer = 0;
pthread_t writer_thread;
int writer_running = 0;
pthread_mutex_t buffer_switch_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t buffer_ready_cond = PTHREAD_COND_INITIALIZER;
pthread_cond_t buffer_written_cond = PTHREAD_COND_INITIALIZER;

// Precomputed lookup tables for faster random generation
CumulativeFreq iub_lookup[15];
CumulativeFreq homosapiens_lookup[4];

// Convert frequency table to cumulative probabilities
void make_cumulative(Frequency *table, int table_size, CumulativeFreq *result) {
    double prob = 0.0;
    for (int i = 0; i < table_size; i++) {
        prob += table[i].p;
        result[i].prob = prob;
        result[i].c = table[i].c;
    }
}

// Binary search to find character based on probability
inline char lookup_char(CumulativeFreq *table, int table_size, double random) {
    // Optimized binary search
    int low = 0;
    int high = table_size - 1;
    
    while (low < high) {
        int mid = (low + high) / 2;
        if (random < table[mid].prob)
            high = mid;
        else
            low = mid + 1;
    }
    
    return table[low].c;
}

// Thread function for asynchronous writing
void* writer_function(void *arg) {
    while (writer_running || buffers[write_buffer].size > 0) {
        pthread_mutex_lock(&buffer_switch_mutex);
        
        // Wait until a buffer is ready
        while (!buffers[write_buffer].ready && writer_running) {
            pthread_cond_wait(&buffer_ready_cond, &buffer_switch_mutex);
        }
        
        // Get the buffer to write
        Buffer *buffer = &buffers[write_buffer];
        size_t size = buffer->size;
        
        // If we're shutting down and there's nothing to write, exit
        if (size == 0 && !writer_running) {
            pthread_mutex_unlock(&buffer_switch_mutex);
            break;
        }
        
        // Mark buffer as not ready
        buffer->ready = 0;
        pthread_mutex_unlock(&buffer_switch_mutex);
        
        // Write the buffer to stdout
        fwrite(buffer->data, 1, size, stdout);
        
        // Update buffer status
        pthread_mutex_lock(&buffer_switch_mutex);
        buffer->size = 0;
        write_buffer = (write_buffer + 1) % NUM_BUFFERS;
        pthread_cond_signal(&buffer_written_cond);
        pthread_mutex_unlock(&buffer_switch_mutex);
    }
    
    return NULL;
}

// Initialize the asynchronous I/O system
void init_io() {
    // Initialize buffers
    for (int i = 0; i < NUM_BUFFERS; i++) {
        buffers[i].size = 0;
        buffers[i].ready = 0;
        pthread_mutex_init(&buffers[i].mutex, NULL);
    }
    
    // Start writer thread
    writer_running = 1;
    pthread_create(&writer_thread, NULL, writer_function, NULL);
}

// Flush current buffer and wait for it to be written
void flush_buffer() {
    pthread_mutex_lock(&buffer_switch_mutex);
    
    // Mark current buffer as ready to be written
    if (buffers[current_buffer].size > 0) {
        buffers[current_buffer].ready = 1;
        pthread_cond_signal(&buffer_ready_cond);
        
        // Switch to next buffer
        current_buffer = (current_buffer + 1) % NUM_BUFFERS;
        
        // Wait until the next buffer is available
        while (buffers[current_buffer].size > 0) {
            pthread_cond_wait(&buffer_written_cond, &buffer_switch_mutex);
        }
    }
    
    pthread_mutex_unlock(&buffer_switch_mutex);
}

// Add data to current buffer, flushing if necessary
void buffer_write(const char *data, size_t size) {
    // If data is too large for buffer, flush and write directly
    if (size > BUFFER_SIZE) {
        flush_buffer();
        fwrite(data, 1, size, stdout);
        return;
    }
    
    // If data won't fit in current buffer, flush it
    if (buffers[current_buffer].size + size > BUFFER_SIZE) {
        flush_buffer();
    }
    
    // Copy data to buffer
    memcpy(buffers[current_buffer].data + buffers[current_buffer].size, data, size);
    buffers[current_buffer].size += size;
}

// Shutdown the asynchronous I/O system
void shutdown_io() {
    // Flush any remaining data
    flush_buffer();
    
    // Signal writer thread to shut down
    pthread_mutex_lock(&buffer_switch_mutex);
    writer_running = 0;
    pthread_cond_signal(&buffer_ready_cond);
    pthread_mutex_unlock(&buffer_switch_mutex);
    
    // Wait for writer thread to finish
    pthread_join(writer_thread, NULL);
    
    // Clean up
    for (int i = 0; i < NUM_BUFFERS; i++) {
        pthread_mutex_destroy(&buffers[i].mutex);
    }
    pthread_mutex_destroy(&buffer_switch_mutex);
    pthread_cond_destroy(&buffer_ready_cond);
    pthread_cond_destroy(&buffer_written_cond);
}

// Generate a repeated sequence
void repeat_fasta(const char *src, int src_len, int n) {
    char buffer[LINE_WIDTH + 1];
    buffer[LINE_WIDTH] = '\n';
    
    int pos = 0;
    int remaining = n;
    
    while (remaining > 0) {
        int line_len = remaining < LINE_WIDTH ? remaining : LINE_WIDTH;
        
        // Optimize for the case where we can copy directly from source
        if (pos + line_len <= src_len) {
            memcpy(buffer, src + pos, line_len);
        } else {
            // Handle the wraparound case
            for (int i = 0; i < line_len; i++) {
                buffer[i] = src[(pos + i) % src_len];
            }
        }
        
        // Add newline if it's a full line
        if (line_len == LINE_WIDTH) {
            buffer_write(buffer, line_len + 1);
        } else {
            // Add newline for partial line
            buffer[line_len] = '\n';
            buffer_write(buffer, line_len + 1);
        }
        
        pos = (pos + line_len) % src_len;
        remaining -= line_len;
    }
}

// Generate a random sequence
double random_fasta(CumulativeFreq *table, int table_size, int n, double seed) {
    char buffer[LINE_WIDTH + 1];
    buffer[LINE_WIDTH] = '\n';
    
    int remaining = n;
    
    while (remaining > 0) {
        int line_len = remaining < LINE_WIDTH ? remaining : LINE_WIDTH;
        
        for (int i = 0; i < line_len; i++) {
            seed = (seed * 3877.0 + 29573.0) % IM;
            buffer[i] = lookup_char(table, table_size, seed / IM);
        }
        
        // Add newline if it's a full line
        if (line_len == LINE_WIDTH) {
            buffer_write(buffer, line_len + 1);
        } else {
            // Add newline for partial line
            buffer[line_len] = '\n';
            buffer_write(buffer, line_len + 1);
        }
        
        remaining -= line_len;
    }
    
    return seed;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <number>\n", argv[0]);
        return 1;
    }
    
    int n = atoi(argv[1]);
    
    // Calculate the size of the frequency tables
    int iub_size = sizeof(iub) / sizeof(iub[0]);
    int homosapiens_size = sizeof(homosapiens) / sizeof(homosapiens[0]);
    
    // Convert the frequency tables to cumulative probabilities
    make_cumulative(iub, iub_size, iub_lookup);
    make_cumulative(homosapiens, homosapiens_size, homosapiens_lookup);
    
    // Initialize I/O system
    init_io();
    
    // Generate the sequences
    const char *header1 = ">ONE Homo sapiens alu\n";
    buffer_write(header1, strlen(header1));
    repeat_fasta(alu, strlen(alu), n * 2);
    
    const char *header2 = ">TWO IUB ambiguity codes\n";
    buffer_write(header2, strlen(header2));
    double seed = 42.0;
    seed = random_fasta(iub_lookup, iub_size, n * 3, seed);
    
    const char *header3 = ">THREE Homo sapiens frequency\n";
    buffer_write(header3, strlen(header3));
    random_fasta(homosapiens_lookup, homosapiens_size, n * 5, seed);
    
    // Shutdown I/O system
    shutdown_io();
    
    return 0;
}

File: fasta/Version3/Claude/fasta3-opt.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// Converted from Python to Rust with optimizations by Claude

use std::env;
use std::io::{self, Write, BufWriter};
use std::sync::{Arc, Mutex};
use std::thread;
use std::collections::VecDeque;
use std::sync::mpsc::{self, Sender, Receiver};

const ALU: &str = concat!(
    "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG",
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA",
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT",
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA",
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG",
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC",
    "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"
);

const WIDTH: usize = 60;
const IM: f64 = 139968.0;
const BUFFER_SIZE: usize = 64 * 1024; // 64KB buffer

// Frequency tables
const IUB: [(&str, f64); 15] = [
    ("a", 0.27), ("c", 0.12), ("g", 0.12), ("t", 0.27),
    ("B", 0.02), ("D", 0.02), ("H", 0.02), ("K", 0.02),
    ("M", 0.02), ("N", 0.02), ("R", 0.02), ("S", 0.02),
    ("V", 0.02), ("W", 0.02), ("Y", 0.02)
];

const HOMOSAPIENS: [(&str, f64); 4] = [
    ("a", 0.3029549426680),
    ("c", 0.1979883004921),
    ("g", 0.1975473066391),
    ("t", 0.3015094502008)
];

// Message type for the writer thread
enum WriterMsg {
    Data(Vec<u8>),
    Done,
}

// Writer thread that handles all output
fn writer_thread(rx: Receiver<WriterMsg>) -> io::Result<()> {
    let stdout = io::stdout();
    let mut writer = BufWriter::with_capacity(BUFFER_SIZE, stdout.lock());
    
    while let Ok(msg) = rx.recv() {
        match msg {
            WriterMsg::Data(data) => {
                writer.write_all(&data)?;
            },
            WriterMsg::Done => {
                break;
            }
        }
    }
    
    writer.flush()?;
    Ok(())
}

// Convert frequency table to cumulative probabilities
fn make_cumulative(table: &[(&str, f64)]) -> (Vec<f64>, Vec<u8>) {
    let mut probs = Vec::with_capacity(table.len());
    let mut chars = Vec::with_capacity(table.len());
    let mut prob = 0.0;
    
    for &(ch, p) in table {
        prob += p;
        probs.push(prob);
        chars.push(ch.as_bytes()[0]);
    }
    
    (probs, chars)
}

// Binary search to find character based on probability - optimized version
#[inline]
fn bisect(probs: &[f64], r: f64) -> usize {
    let mut low = 0;
    let mut high = probs.len() - 1;
    
    while low < high {
        let mid = (low + high) / 2;
        if r < probs[mid] {
            high = mid;
        } else {
            low = mid + 1;
        }
    }
    
    low
}

// Generate a repeated sequence
fn repeat_fasta(tx: &Sender<WriterMsg>, src: &str, n: usize) -> io::Result<()> {
    let src_bytes = src.as_bytes();
    let src_len = src.len();
    
    // Create an expanded version of the source to reduce modulo operations
    let expanded_src = {
        let mut expanded = Vec::with_capacity(src_len * 2);
        expanded.extend_from_slice(src_bytes);
        expanded.extend_from_slice(src_bytes);
        expanded
    };
    
    let mut pos = 0;
    let mut remaining = n;
    
    // Use a buffer to reduce the number of channel sends
    let mut buffer = Vec::with_capacity(BUFFER_SIZE);
    let mut line = vec![0; WIDTH + 1];
    line[WIDTH] = b'\n';
    
    while remaining > 0 {
        let line_len = remaining.min(WIDTH);
        
        // Copy from the expanded source to avoid modulo operations
        if pos + line_len <= expanded_src.len() {
            line[..line_len].copy_from_slice(&expanded_src[pos..pos + line_len]);
        } else {
            // Fall back to modulo for the rare case where we need to wrap
            for i in 0..line_len {
                line[i] = src_bytes[(pos + i) % src_len];
            }
        }
        
        // Add to buffer
        if line_len == WIDTH {
            buffer.extend_from_slice(&line[..WIDTH + 1]);
        } else {
            line[line_len] = b'\n';
            buffer.extend_from_slice(&line[..line_len + 1]);
        }
        
        // Send buffer when it's full enough
        if buffer.len() >= BUFFER_SIZE - (WIDTH + 1) {
            tx.send(WriterMsg::Data(buffer)).unwrap();
            buffer = Vec::with_capacity(BUFFER_SIZE);
        }
        
        pos = (pos + line_len) % src_len;
        remaining -= line_len;
    }
    
    // Send any remaining data
    if !buffer.is_empty() {
        tx.send(WriterMsg::Data(buffer)).unwrap();
    }
    
    Ok(())
}

// Generate a random sequence
fn random_fasta(
    tx: &Sender<WriterMsg>,
    probs: &[f64],
    chars: &[u8],
    n: usize,
    mut seed: f64
) -> io::Result<f64> {
    let mut line = vec![0; WIDTH + 1];
    line[WIDTH] = b'\n';
    
    let mut buffer = Vec::with_capacity(BUFFER_SIZE);
    let mut remaining = n;
    
    // Pre-generate a table of characters for common probability ranges
    // This is a simple form of "fast path" optimization
    let char_table = if chars.len() <= 4 {  // Only for small tables like homosapiens
        let mut table = vec![0u8; 1024];
        for i in 0..1024 {
            let r = i as f64 / 1024.0;
            let index = bisect(probs, r);
            table[i] = chars[index];
        }
        Some(table)
    } else {
        None
    };
    
    while remaining > 0 {
        let line_len = remaining.min(WIDTH);
        
        for i in 0..line_len {
            // Generate random number
            seed = (seed * 3877.0 + 29573.0) % IM;
            let r = seed / IM;
            
            // Find character based on probability
            line[i] = if let Some(table) = &char_table {
                // Fast path for common cases
                let table_index = (r * 1024.0) as usize;
                if table_index < 1024 {
                    table[table_index]
                } else {
                    chars[bisect(probs, r)]
                }
            } else {
                chars[bisect(probs, r)]
            };
        }
        
        // Add to buffer
        if line_len == WIDTH {
            buffer.extend_from_slice(&line[..WIDTH + 1]);
        } else {
            line[line_len] = b'\n';
            buffer.extend_from_slice(&line[..line_len + 1]);
        }
        
        // Send buffer when it's full enough
        if buffer.len() >= BUFFER_SIZE - (WIDTH + 1) {
            tx.send(WriterMsg::Data(buffer)).unwrap();
            buffer = Vec::with_capacity(BUFFER_SIZE);
        }
        
        remaining -= line_len;
    }
    
    // Send any remaining data
    if !buffer.is_empty() {
        tx.send(WriterMsg::Data(buffer)).unwrap();
    }
    
    Ok(seed)
}

fn main() -> io::Result<()> {
    // Parse command-line argument
    let n = env::args()
        .nth(1)
        .unwrap_or_else(|| "1000".to_string())
        .parse::<usize>()
        .expect("Expected a number");
    
    // Create channel for writer thread
    let (tx, rx) = mpsc::channel();
    
    // Start writer thread
    let writer_handle = thread::spawn(move || {
        writer_thread(rx)
    });
    
    // Create cumulative frequency tables
    let (iub_probs, iub_chars) = make_cumulative(&IUB);
    let (homosapiens_probs, homosapiens_chars) = make_cumulative(&HOMOSAPIENS);
    
    // Generate sequences
    let header1 = b">ONE Homo sapiens alu\n";
    tx.send(WriterMsg::Data(header1.to_vec())).unwrap();
    repeat_fasta(&tx, ALU, n * 2)?;
    
    let header2 = b">TWO IUB ambiguity codes\n";
    tx.send(WriterMsg::Data(header2.to_vec())).unwrap();
    let seed = random_fasta(&tx, &iub_probs, &iub_chars, n * 3, 42.0)?;
    
    let header3 = b">THREE Homo sapiens frequency\n";
    tx.send(WriterMsg::Data(header3.to_vec())).unwrap();
    random_fasta(&tx, &homosapiens_probs, &homosapiens_chars, n * 5, seed)?;
    
    // Signal that we're done
    tx.send(WriterMsg::Done).unwrap();
    
    // Wait for writer thread to finish
    writer_handle.join().unwrap()?;
    
    Ok(())
}